cpu/esp32/include/periph_cpu_esp32s2.h

```diff
@@ -0,0 +1,227 @@
+ * purpose SPI:
+ *
+ * - controller SPI0 is reserved for external memories like Flash and PSRAM
+ * - controller SPI1 is reserved for external memories like Flash and PSRAM
+ * - controller SPI2 can be used for peripherals (also called FSPI)
+ * - controller SPI3 can be used for peripherals
+ *
+ * Thus, SPI2 (`FSPI`) and SPI3 can be used as general purpose SPI in
+ * RIOT as SPI_DEV(0) and SPI_DEV(1) by defining the symbols `SPI0_*`
+ * and `SPI1_*`.
+ *
+ * The following table shows the pin configuration used by default, even
+ * though it **can vary** from board to board.
+ *
+ * <center>
+ *
+ * Device (Alias)          | Signal | Pin    | Symbol      | Remarks
+ * :-----------------------|:------:|:-------|:-----------:|:---------------------------
+ * `SPI_HOST0`/`SPI_HOST1` | SPICS0 | GPIO29 | - | reserved for Flash and PSRAM
+ * `SPI_HOST0`/`SPI_HOST1` | SPICS1 | GPIO26 | - | reserved for Flash and PSRAM
+ * `SPI_HOST0`/`SPI_HOST1` | SPICLK | GPIO30 | - | reserved for Flash and PSRAM
+ * `SPI_HOST0`/`SPI_HOST1` | SPID   | GPIO32 | - | reserved for Flash and PSRAM
+ * `SPI_HOST0`/`SPI_HOST1` | SPIQ   | GPIO31 | - | reserved for Flash and PSRAM
+ * `SPI_HOST0`/`SPI_HOST1` | SPIHD  | GPIO27 | - | reserved for Flash and PSRAM (only in `qio` or `qout` mode)
+ * `SPI_HOST0`/`SPI_HOST1` | SPIWP  | GPIO28 | - | reserved for Flash and PSRAM (only in `qio` or `qout` mode)
+ * `SPI_HOST0`/`SPI_HOST1` | SPIIO4 | GPIO33 | - | reserved for Flash and PSRAM (only in octal mode)
+ * `SPI_HOST0`/`SPI_HOST1` | SPIIO5 | GPIO34 | - | reserved for Flash and PSRAM (only in octal mode)
+ * `SPI_HOST0`/`SPI_HOST1` | SPIIO6 | GPIO35 | - | reserved for Flash and PSRAM (only in octal mode)
+ * `SPI_HOST0`/`SPI_HOST1` | SPIIO7 | GPIO36 | - | reserved for Flash and PSRAM (only in octal mode)
+ * `SPI_HOST0`/`SPI_HOST1` | SPIDQA | GPIO37 | - | reserved for Flash and PSRAM (only in octal mode)
+ * `SPI_HOST2` (`FSPI`)    | SCK    | GPIO36 |`#SPI0_SCK`  | can be used
+ * `SPI_HOST2` (`FSPI`)    | MOSI   | GPIO35 |`#SPI0_MOSI` | can be used
+ * `SPI_HOST2` (`FSPI`)    | MISO   | GPIO37 |`#SPI0_MISO` | can be used
+ * `SPI_HOST2` (`FSPI`)    | CS0    | GPIO38 |`#SPI0_CS0`  | can be used
+ * `SPI_HOST3` (`HSPI`)    | SCK    | -      |`#SPI1_SCK`  | can be used
+ * `SPI_HOST3` (`HSPI`)    | MOSI   | -      |`#SPI1_MOSI` | can be used
+ * `SPI_HOST3` (`HSPI`)    | MISO   | -      |`#SPI1_MISO` | can be used
+ * `SPI_HOST3` (`HSPI`)    | CS0    | -      |`#SPI1_CS0`  | can be used
+ *
+ * </center><br>
+ */
+
+/**
+ * @name   Timer configuration depending on which implementation is used
+ *
+ * Timers are MCU built-in feature and not board-specific. They are therefore
+ * configured here.
+ *
+ * ESP32-S2 has two timer groups with two timers each, resulting in a total of
+ * four timers. Since one timer is used as system timer, up to three timers
+ * with one channel each can be used in RIOT as timer devices
+ * TIMER_DEV(0) ... TIMER_DEV(2).
```

Comment:
Shouldn't `TIMER_NUMOF` be 3 then?

Reply:
If the hardware counter implementation for `periph/timer` is used (`MODULE_ESP_HW_COUNTER` is defined), there are three timer devices, but the third timer device uses a high-level interrupt that cannot be disabled. Therefore, the number of timer devices is limited to two, so only `TIMER_DEV(0)` and `TIMER_DEV(1)` can be used.

In case of the hardware timer implementation for `periph/timer` (`MODULE_ESP_HW_COUNTER` is not defined), the number of timers `TIMER_NUMOF` is derived from `SOC_TIMER_GROUP_TOTAL_TIMERS` here https://github.com/RIOT-OS/RIOT/blob/c908a60b7f03af55abc5e3337520fce633d318d6/cpu/esp32/include/periph_cpu.h#L659.